fluid/conv_seq_to_seq/model.py

```diff
@@ -0,0 +1,238 @@
+            x = fluid.nets.glu(x, dim=2)
+            x = (x + residual) * math.sqrt(0.5)
+
+        # # T x B x C -> B x T x C
+        # x = Op.transpose(x, 1, 2)
+
+        # project back to size of embedding
+        x = self.fc2(x)
+
+        y = (x + input_embedding) * math.sqrt(0.5)
+
+        return x, y
+
+
+class ConvDecoder:
+    '''
+    decode trainer
+    '''
+
+    def __init__(self,
+                 dict_size,
+                 embed_dim,
+                 out_embed_dim,
+                 max_positions,
+                 convolutions,
+                 pad_id,
+                 pos_pad_id,
+                 attention=True,
+                 dropout=0.1,
+                 share_embed=False):
+        self.dropout = dropout
+
+        in_channels = convolutions[0][0]
+        if isinstance(attention, bool):
+            attention = [attention] * len(convolutions)
+
+        self.embed_tokens = Embedding(dict_size + 1, embed_dim, pad_id)
+        self.embed_positions = Embedding(
+            max_positions + 1,
+            embed_dim,
+            pos_pad_id, )
+
+        self.fc1 = Linear(in_channels, dropout=dropout)
+        self.projections = []
+        self.convolutions = []
+        self.attention = []
+        for i, (out_channels, kernel_size) in enumerate(convolutions):
+            pad = kernel_size - 1
+            self.projections.append(
+                Linear(out_channels) if in_channels != out_channels else None)
+            self.convolutions.append(
+                Conv1D(
+                    in_channels,
+                    out_channels * 2,
+                    kernel_size,
+                    # padding=pad,
+                    dropout=dropout))
+            self.attention.append(
+                AttentionLayer(out_channels, embed_dim)
+                if attention[i] else None)
+            in_channels = out_channels
+
+        self.fc2 = Linear(out_embed_dim)
+        self.fc3 = Linear(dict_size + 1, dropout=dropout)
+
+    def forward(self, prev_output_tokens, prev_positions, encoder_out):
+        '''
+        only works in train mode
+        '''
+        encoder_a, encoder_b = self._split_encoder_out(encoder_out)
+
+        x = self.embed_tokens(prev_output_tokens)
+        x += self.embed_positions(prev_positions)
+        x = Op.dropout(x, self.dropout)
+        target_embedding = x
+
+        # project to size of convolution
+        x = self.fc1(x)
+
+        x = self._transpose_if_training(x)
+
+        # temporal covolutoins
+        avg_attn_scores = None
+        num_attn_layers = len(self.attention)
+        for proj, conv, attention in zip(self.projections, self.convolutions,
+                                         self.attention):
+            residual = x if proj is None else proj(x)
+            x = Op.dropout(x, self.dropout)
+            x = conv(x)
+            x = fluid.nets.glu(x, dim=2)
+
+            if attention is not None:
+                x = self._transpose_if_training(x)
```

Comment:
I think this is also used when the shape of input `x` is `T x B x C` and to convert the shape to `B x T x C` for attention calculations. 
Maybe we can try to remove `_transpose_if_training` and  `_split_encoder_out ` included `transpose` and unify the shapes to `B x T x C` temporarily if we can confirm that our here wrapped `Conv1D` is not suitable to `T x B x C`. 
I am not sure the reason why fairseq-py uses `T x B x C` shape for convolution, maybe it is faster according to the [annotations](https://github.com/facebookresearch/fairseq-py/blob/master/fairseq/modules/conv_tbc.py#L17).